subsys/net/ip/net_tls.c

```diff
@@ -0,0 +1,735 @@
+	sec_tag_t tag;
+	const void *buf;
+	size_t len;
+};
+
+/* Global poll of credentials shared among TLS contexts. */
+static struct net_tls_credential credentials[CONFIG_NET_MAX_CREDENTIALS_NUMBER];
+
+/* A poll of TLS contexts. */
+static struct net_tls tls_context[CONFIG_NET_TLS_MAX_CONTEXT];
+
+#if defined(MBEDTLS_DEBUG_C) && defined(CONFIG_NET_TLS_DEBUG)
+static void my_debug(void *ctx, int level,
+		     const char *file, int line, const char *str)
+{
+	const char *p, *basename;
+	int len;
+
+	ARG_UNUSED(ctx);
+
+	/* Extract basename from file */
+	for (p = basename = file; *p != '\0'; p++) {
+		if (*p == '/' || *p == '\\') {
+			basename = p + 1;
+		}
+
+	}
+
+	/* Avoid printing double newlines */
+	len = strlen(str);
+	if (str[len - 1] == '\n') {
+		((char *)str)[len - 1] = '\0';
+	}
+
+	NET_DBG("%s:%04d: |%d| %s", basename, line, level, str);
+}
+#endif /* defined(MBEDTLS_DEBUG_C) && defined(CONFIG_NET_TLS_DEBUG) */
+
+static int tls_entropy_func(void *ctx, unsigned char *buf, size_t len)
+{
+	/* TODO Provide a decent entropy source. */
+
+	ARG_UNUSED(ctx);
+
+	size_t i = len / 4;
+	u32_t val;
+
+	while (i--) {
+		val = sys_rand32_get();
+		UNALIGNED_PUT(val, (u32_t *)buf);
+		buf += 4;
+	}
+
+	i = len & 0x3;
+	val = sys_rand32_get();
+	while (i--) {
+		*buf++ = val;
+		val >>= 8;
+	}
+
+	return 0;
+}
+
+static struct net_tls_credential *unused_credential_get(void)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(credentials); i++) {
+		if (credentials[i].type == NET_TLS_CREDENTIAL_UNUSED) {
+			return &credentials[i];
+		}
+	}
+
+	return NULL;
+}
+
+static struct net_tls_credential *credential_get(
+		sec_tag_t tag, enum net_tls_credential_type type)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(credentials); i++) {
+		if (credentials[i].type == type && credentials[i].tag == tag) {
+			return &credentials[i];
+		}
+	}
+
+	return NULL;
+}
+
+static struct net_tls_credential *credential_next_get(
+		sec_tag_t tag, struct net_tls_credential *iter)
+{
+	int i;
+
+	if (!iter) {
+		iter = credentials;
+	} else {
+		iter++;
+	}
+
+	for (i = iter - credentials; i < ARRAY_SIZE(credentials); i++) {
+		if (credentials[i].type != NET_TLS_CREDENTIAL_UNUSED &&
+		    credentials[i].tag == tag) {
+			return &credentials[i];
+		}
+	}
+
+	return NULL;
+}
+
+static int tls_tx(void *ctx, const unsigned char *buf, size_t len)
+{
+	struct net_context *context = ctx;
+	struct net_pkt *pkt;
+	int bytes, ret;
+
+	pkt = net_pkt_get_tx(context, TIMEOUT_TLS_TX_MS);
+
+	if (!pkt) {
+		return -EIO;
+	}
+
+	bytes = net_pkt_append(pkt, len, (u8_t *)buf, TIMEOUT_TLS_TX_MS);
+
+	ret = net_context_output(context, pkt, &context->remote);
+	if (ret < 0) {
+		net_pkt_unref(pkt);
+		return ret;
+	}
+
+	return bytes;
+}
+
+static int tls_rx(void *ctx, unsigned char *buf, size_t len)
+{
+	struct net_context *context = ctx;
+	int ret;
+	u16_t len_left;
+
+	if (!context->tls->rx_pkt) {
+		context->tls->rx_pkt =
+			k_fifo_get(&context->tls->rx_fifo, K_NO_WAIT);
```

Comment:
Nice, so due to layering violation, this now duplicates the code from lib/sockets/sockets.c. Really nice... ;-)